feat: add ContractType.dev_messages support

helloibis opened the pull request:
### What I did

fixes: APE-308

Requires https://github.com/ApeWorX/ethpm-types/pull/51
Facilitates https://github.com/ApeWorX/ape/pull/1125

### How I did it

During the `compile()` step we parse the source line-by-line and save the line number + message if we encounter a dev message (`#dev: foobar`)

### How to verify it

Via `test_compile_parse_dev_messages()` and the new `contract_with_dev_messages.vy` test contract.

### Checklist

- [x] Passes all linting checks (pre-commit and CI jobs)
- [x] New test cases have been added and are passing
- [x] Documentation has been updated
- [x] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

helloibis commented:
Since `ape-vyper` doesn't explicitly depend on ethpm-types, is this how we pull this off?

1. Merge `ethpm-types` PR and make a new version
2. Merge `ape` PR and create the alpha version (we should hold on this until after Monday so it doesn't affect/block releasing 0.5.5)
3. Update this PR so `setup.py` uses the `0.6.0a` alpha release of `ape`
4. Merge this PR and publish an `ape-vyper` alpha version `0.6.0a`

cc @unparalleled-js @NotPeopling2day @fubuloubu

unparalleled-js commented on ape_vyper/compiler.py:
```diff
@@ -137,6 +136,14 @@ def compile(
+
+                dev_messages = {}
+                for line_num, line in enumerate(source.split("\n")):
+                    if match := re.search(r"#\s*dev\s*:\s*(.+)", line):
```
As an optimization, we could compile this regex pattern outside the for loop (or once globally at top of module)

```python
dev_str_pattern = re.compile(r"#\s*dev\s*:\s*(.+)")
for line_num, line in enumerate(source.split("\n")):
    if match := re.search(dev_str_pattern, line):
```

helloibis replied:
Good point! Updated in a2b1f01

unparalleled-js commented on ape_vyper/compiler.py:
```diff
@@ -137,6 +136,14 @@ def compile(
                 result["pcmap"] = result["source_map"]["pc_pos_map"]
+
+                dev_messages = {}
+                for line_num, line in enumerate(source.split("\n")):
```
is this actually more like `line_index` and `line_index + 1` is the line number (like we do below)?
Asking because wondering if should rename this local variable to `line_index` to not confuse with actual `line_number`.

helloibis replied:
Yeah true, I think that name would be much better so it's clear that the resulting map key will be +1 since those aren't zero-indexed

Updated in a2b1f01

unparalleled-js reviewed (approved):
If this requires a specific `ethpm_types` version (one that is higher than Core ape), I think it is okay to put that constraint in its `setup.py`.

helloibis commented:
> If this requires a specific ethpm_types version (one that is higher than Core ape)

@unparalleled-js good point, after the `ethpm-types` release I'll add the dependency to this `setup.py`

helloibis commented:
Context for fcc1f33

So I realized Brownie includes the `dev:` portion when parsing out dev messages, to keep parity with that I changed the regex to include that as well. I updated the tests to reflect this and removed the "weird spacing" test. So basically the form is just:

`#\s*(dev:.*)`

Which keeps things a lot simpler!
